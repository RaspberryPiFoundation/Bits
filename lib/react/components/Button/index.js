var _jsxFileName = 'src/components/Button/index.js',
    _this = this;

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import React from 'react';

var Button = function Button(_ref) {
  var props = _objectWithoutProperties(_ref, []);

  return React.createElement(
    'button',
    Object.assign({}, props, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 4
      },
      __self: _this
    }),
    children
  );
};

export default Button;