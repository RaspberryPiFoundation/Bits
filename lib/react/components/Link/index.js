var _jsxFileName = 'src/components/Link/index.js',
    _this = this;

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { Link as RouterLink } from 'react-router-dom';
// import './Link.scss'

var Link = function Link(_ref) {
  var className = _ref.className,
      to = _ref.to,
      props = _objectWithoutProperties(_ref, ['className', 'to']);

  var classes = classnames('c-link', className);

  return React.createElement(
    RouterLink,
    Object.assign({ className: classes, to: to }, props, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 11
      },
      __self: _this
    }),
    children
  );
};

Link.propTypes = {
  className: PropTypes.string,
  to: PropTypes.string.isRequired
};

export default Link;