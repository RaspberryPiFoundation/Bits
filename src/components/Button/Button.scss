@import '../../styles/component-helper';

$button-border-radius:              $border-radius;
$button-border-width:               $border-width;
$button-font-weight:                map-get($type-font-weights, sub-heading);
$button-line-height:                $baseline-x7;
$button-line-height-md:             $baseline-x8;
$button-ph:                         $baseline-x4;
$button-type-map:                   (
  null: ($type-sub-heading-font-size,    $button-line-height),
  md:   ($type-sub-heading-font-size-md, $button-line-height-md),
);

$button-primary-background-color:   $color-ui-yellow;
$button-primary-border-color:       $color-ui-yellow;
$button-primary-color:              $color-ui-greyscale-darkest;

$button-secondary-background-color: $color-white;
$button-secondary-border-color:     $color-ui-greyscale-mid;
$button-secondary-color:            $color-ui-greyscale-mid;

.c-button {
  @include iota-type($button-type-map);

  border-radius:   $button-border-radius;
  border-style:    solid;
  border-width:    $button-border-width;
  cursor:          pointer;
  display:         inline-block;
  font-weight:     $button-font-weight;
  padding:         0 $button-ph;
  text-align:      center;
  vertical-align:  middle;
  text-decoration: none;
}

.c-button--primary {
  background-color: $button-primary-background-color;
  border-color:     $button-primary-border-color;
  color:            $button-primary-color;
}

.c-button--secondary {
  background-color: $button-secondary-background-color;
  border-color:     $button-secondary-border-color;
  color:            $button-secondary-color;
}

.c-button--block {
  display: block;
  width:   100%;
}
