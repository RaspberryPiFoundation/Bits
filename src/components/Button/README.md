## Primary Button

```js
<Button>Click me!</Button>
```

### Variants

#### Secondary Button

```js
<Button secondary>Secondary button</Button>
```

### Light-on-Dark

```js
<LightOnDark uBgcolor="ui-black">
  <Button secondary lightOnDark>Use this one on dark backgrounds</Button>
</LightOnDark>
```

## Block

```js
<Button block>Block Button</Button>
```
