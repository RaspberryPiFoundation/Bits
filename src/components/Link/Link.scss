@import '../../styles/component-helper';

.c-link {
  @include link;
}

.c-link--light-on-dark {
  @include link-light-on-dark;
}
