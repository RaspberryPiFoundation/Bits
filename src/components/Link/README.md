```jsx
import { MemoryRouter } from 'react-router-dom';
<MemoryRouter>
  <Link to="https://www.raspberrypi.org">Link text goes here</Link>
</MemoryRouter>
```

#### In HTML/CSS

```html static
<a class="c-link" href="https://www.raspberrypi.org">Link text goes here</a>
```
