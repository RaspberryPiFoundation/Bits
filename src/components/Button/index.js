import { Base } from 'react-iotacss'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import './Button.scss'
import Link from '../Link'

/**
 * A styled button that can also be used as a link
 */
const Button = ({
  children,
  className,
  lightOnDark,
  secondary,
  to,
  ...props
}) => {
  const classNames = classnames('c-button', className, {
    'c-button--light-on-dark': lightOnDark,
    'c-button--secondary': secondary,
  })
  const isLink = () => typeof to === 'string'

  if (isLink()) {
    return (
      <Link {...props} className={classNames} excludeClassname={true} to={to}>
        {children}
      </Link>
    )
  }

  return (
    <Base {...props} className={classNames} tagName="button">
      {children}
    </Base>
  )
}

Button.propTypes = {
  /** Any additional classes you wish to add to the button */
  className: PropTypes.string,
  /** The contents of the button */
  children: PropTypes.node.isRequired,
  /** Changes the hover styles to better suit dark backgrounds */
  lightOnDark: PropTypes.bool,
  /** Applies secondary button styling */
  secondary: PropTypes.bool,
  /** Causes the button to be treated as a link to the URL specified */
  to: PropTypes.string,
}

export default Button
