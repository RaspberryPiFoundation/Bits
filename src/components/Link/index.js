import { Base } from 'react-iotacss'
import { Link as RouterLink } from 'react-router-dom'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import './Link.scss'

const Link = ({ children, className, excludeClassname, to, ...props }) => {
  const classNames = classnames(
    {
      'c-link': !excludeClassname,
    },
    className,
  )

  const anchorLink = (
    <Base className={classNames} href={to} tagName="a" {...props}>
      {children}
    </Base>
  )

  const isExternal = () => !/^https?:\/\//i.test(to)

  const routerLink = (
    <RouterLink className={classNames} to={to} {...props}>
      {children}
    </RouterLink>
  )

  if (isExternal()) {
    return anchorLink
  }

  return routerLink
}

Link.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  excludeClassname: PropTypes.bool,
  to: PropTypes.string.isRequired,
}

export default Link
