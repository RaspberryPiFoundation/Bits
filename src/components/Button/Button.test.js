import { mount, shallow } from 'enzyme'
import Button from '../Button'
import { MemoryRouter } from 'react-router-dom'
import Link from '../Link'
import React from 'react'

const children = 'Whatever'

let button

describe('<Button />', () => {
  describe('Renders...', () => {
    beforeEach(() => {
      button = shallow(<Button>{children}</Button>)
    })

    it('without crashing', () => {
      expect(button.find('.c-button')).toBePresent()
    })

    it('matches its snapshot', () => {
      expect(button).toMatchSnapshot()
    })

    describe('external links', () => {
      it('renders a Link component', () => {
        let externalButton = mount(
          <MemoryRouter>
            <Button to="https://www.100yen.co.uk">{children}</Button>
          </MemoryRouter>,
        )
        expect(externalButton.find(Link)).toBePresent()
      })
    })
  })

  describe('Modifiers', () => {
    describe('block', () => {
      it('adds the correct className', () => {
        button = shallow(<Button block>{children}</Button>)

        expect(button.find('.c-button')).toHaveClassName('c-button--block')
      })
    })

    describe('lightOnDark', () => {
      it('adds the correct className', () => {
        button = shallow(<Button lightOnDark>{children}</Button>)

        expect(button.find('.c-button')).toHaveClassName(
          'c-button--light-on-dark',
        )
      })
    })

    describe('secondary', () => {
      it('adds the correct className', () => {
        button = shallow(<Button secondary>{children}</Button>)

        expect(button.find('.c-button')).toHaveClassName('c-button--secondary')
      })
    })
  })
})
