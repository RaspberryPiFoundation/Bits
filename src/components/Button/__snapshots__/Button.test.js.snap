// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Renders... matches its snapshot 1`] = `
<Base
  className="c-button"
  tagName="button"
>
  Whatever
</Base>
`;
